java.base/src/main/java/java/lang/Class$_native.java

```diff
@@ -50,11 +50,23 @@ private static void registerNatives() {
     }
 
     private static Class<?> forName0(String name, boolean initialize, ClassLoader loader, Class<?> caller) throws ClassNotFoundException {
-        Class<?> cls = VMHelpers.findLoadedClass(name, loader);
-        if (cls == null) {
-            throw new ClassNotFoundException(name);
+        // Only classes that were loaded and reachable at build time can be found at runtime.
+
+        // (a) Walk up classloader chain
```

Comment:
A given class loader is not required to delegate to its parent - and IIRC the class loaders in Quarkus sometimes do not. Now that things are a little farther along, this should probably delegate to `loader.loadClass(name)` or similar (there is probably some extra stuff that has to be done to convert `caller` into a reference to the module doing the loading).

WDYT?

Reply:
Taking a look at doing something closer to being correct....one can fell the weight of 25 years of history tracing through the code path taken by forName0 😜

Reply:
Not complete, but closer to correct and still handles the forName in Quarkus bootstrapping that uses the app class loader that I'm working on...